pkg/labels/http_applicator.go

```diff
@@ -0,0 +1,84 @@
+package labels
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/url"
+
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/util"
+)
+
+type httpApplicator struct {
+	client *http.Client
+	// The endpoint that will be queried for matches.
+	// GetMatches will add to this endpoint a query parameter with key "selector" and value selector.String()
+	matchesEndpoint *url.URL
+	logger          logging.Logger
+}
+
+var _ Applicator = &httpApplicator{}
+
+func NewHttpApplicator(client *http.Client, matchesEndpoint *url.URL) (*httpApplicator, error) {
+	if matchesEndpoint == nil {
+		return nil, util.Errorf("matches endpoint cannot be nil")
+	}
+
+	c := client
+	if c == nil {
+		c = http.DefaultClient
+	}
+
+	return &httpApplicator{
+		logger:          logging.DefaultLogger,
+		matchesEndpoint: matchesEndpoint,
+		client:          c,
+	}, nil
+}
+
+func (h *httpApplicator) SetLabel(labelType Type, id, name, value string) error {
+	return util.Errorf("SetLabel not implemented for HttpApplicator (type %s, id %s, name %s, value %s)", labelType, id, name, value)
+}
+
+func (h *httpApplicator) RemoveLabel(labelType Type, id, name string) error {
+	return util.Errorf("RemoveLabel not implemented for HttpApplicator (type %s, id %s, name %s)", labelType, id, name)
+}
+
+func (h *httpApplicator) GetLabels(labelType Type, id string) (Labeled, error) {
+	return Labeled{}, util.Errorf("GetLabels not implemented for HttpApplicator (type %s, id %s)", labelType, id)
+}
+
+func (h *httpApplicator) GetMatches(selector Selector, labelType Type) ([]Labeled, error) {
```

Comment:
Given the current URL structure of our backend, I guess that would require some changes to this code (because when querying for node matches we include "node" in the URL rather than, say, "type=node" in query parameters). So I envision something like a map[Type]string indicating the path to query for a given Type. We can easily make that happen when the time comes.

I have just added some minor help to backends that support both NODE and POD by including "type=node" or "type=pod" in the query parameters since that was super easy, but the existing backend doesn't support it (it doesn't hurt it either, of course, otherwise I wouldn't add it).

The only current user we have of httpApplicator doesn't care (it only needs to look for nodes not pods), so I'm going to ship it as-is.